server/s3_copy_object_action.cc

```diff
@@ -234,74 +223,214 @@ void S3CopyObjectAction::validate_copyobject_request() {
 }
 
-// read source object
-void S3CopyObjectAction::read_object() {
+// Copy source object to destination object
+void S3CopyObjectAction::copy_object() {
   s3_log(S3_LOG_INFO, request_id, "Entering\n");
-  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
-}
 
-// write to destination object
-void S3CopyObjectAction::initiate_data_streaming() {
-  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  if (!total_data_to_stream) {
+    s3_log(S3_LOG_DEBUG, request_id, "Source object is empty");
+    next();
+    return;
+  }
+  motr_reader = motr_reader_factory->create_motr_reader(
+      request, source_object_metadata->get_oid(),
+      source_object_metadata->get_layout_id());
+  motr_reader->set_last_index(0);
+
+  bytes_left_to_read = total_data_to_stream;
+  read_data_block();
+
   s3_log(S3_LOG_DEBUG, "", "Exiting\n");
 }
 
-// Destination bucket
-void S3CopyObjectAction::fetch_bucket_info_failed() {
-  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+void S3CopyObjectAction::read_data_block() {
+  s3_log(S3_LOG_INFO, request_id, "Entering");
 
-  s3_copy_action_state = S3CopyObjectActionState::validationFailed;
+  assert(!read_in_progress);
+  assert(bytes_left_to_read > 0);
 
-  if (bucket_metadata->get_state() == S3BucketMetadataState::missing) {
-    s3_log(S3_LOG_DEBUG, request_id, "Bucket not found\n");
-    set_s3_error("NoSuchBucket");
-  } else if (bucket_metadata->get_state() ==
-             S3BucketMetadataState::failed_to_launch) {
-    s3_log(S3_LOG_ERROR, request_id,
-           "Bucket metadata load operation failed due to pre launch failure\n");
+  const size_t max_blocks_in_one_read_op =
+      S3Option::get_instance()->get_motr_units_per_request();
+  const size_t rest_blocks =
+      (bytes_left_to_read + motr_unit_size - 1) / motr_unit_size;
+
+  if (motr_reader->read_object_data(
+          std::min(rest_blocks, max_blocks_in_one_read_op),
+          std::bind(&S3CopyObjectAction::read_data_block_success, this),
+          std::bind(&S3CopyObjectAction::read_data_block_failed, this))) {
+
+    s3_log(S3_LOG_DEBUG, request_id, "Data read is launched");
+    read_in_progress = true;
+    return;
+  }
+  if (motr_reader->get_state() == S3MotrReaderOpState::failed_to_launch) {
     set_s3_error("ServiceUnavailable");
+    s3_log(S3_LOG_ERROR, request_id,
+           "read_object_data() failed due to motr_entity_open failure\n");
   } else {
-    s3_log(S3_LOG_DEBUG, request_id, "Bucket metadata fetch failed\n");
     set_s3_error("InternalError");
   }
   send_response_to_s3_client();
-  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+
+  s3_log(S3_LOG_DEBUG, NULL, "Exiting\n");
 }
 
-// Destination object
-void S3CopyObjectAction::fetch_object_info_failed() {
-  s3_log(S3_LOG_INFO, request_id, "Entering\n");
-  next();
-  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+void S3CopyObjectAction::read_data_block_success() {
+  assert(read_in_progress);
+
+  s3_log(S3_LOG_INFO, request_id, "Read object data succeeded");
```

Comment:
this log information is misleading... we may have not read whole object. Can we log x bytes read from total_bytes ?

Reply:
Agree. The message is misleading. Will fix.
Will add number of bytes been read below